pkg/artifactory/repositories.go

```diff
@@ -671,6 +673,24 @@ func unpackBaseRemoteRepo(s *schema.ResourceData, packageType string) RemoteRepo
 	return repo
 }
 
+// Special handling of DefaultDeploymentRepo field
+// Artifactory REST API will not accept empty string or null to reset value to not set
+// Instead, using a non-existant repo key works as a workaround
+// To ensure we don't accidentally set the value to a valid repo, we use a UUID v4 string
+// for the repo key
+func getDefaultDeploymentRepo(d *ResourceData) string {
+	key := "default_deployment_repo"
+	value := d.getString(key, false)
+
+	// When value has changed and is empty string, then it has been removed from
+	// the Terraform configuration.
+	if value == "" && d.HasChange(key) {
```

Comment:
Actually, any randomness could do. You could just do `fmt.Sprintf("%d", randInt())` where randInt() is in the utils code. Then you don't need a new dependency

Reply:
Actually, wait a minute!

Look at [./pkg/artifactory/repositories.go](repositories.go):

```go
	"description": {
		Type:     schema.TypeString,
		Optional: true,
		Computed: true,
		DiffSuppressFunc: func(_, old, new string, _ *schema.ResourceData) bool {
			// this is literally what comes back from the server
			return old == fmt.Sprintf("%s (local file cache)", new)
		},
	},
```
or [./pkg/artifactory/resource_artifactory_user.go](resource_artifactory_user.go):
```go
				StateFunc: func(str interface{}) string {
					// Avoid storing the actual value in the state and instead store the hash of it
					value, ok := str.(string)
					if !ok {
						panic(fmt.Errorf("'str' is not a string %s", str))
					}
					hash := sha256.Sum256([]byte(value))
					return base64.StdEncoding.EncodeToString(hash[:])
				},
```

You may want to set a `StateFunc`, which gets called before the diff, that actually has this logic: 

```go
if value == "" {
   return fmt.Sprintf("non-existant-repo-%d", randomInt())
}
return value
```

Then the original field `getString` should work as-is as this will set the local state to the above, and that's what will get sent over-the-wire

Reply:
@chb0github `StateFunc` returns the value to be stored in TF state (according to https://www.terraform.io/docs/extend/schemas/schema-behaviors.html), which is opposite of what I'm trying to achieve here. The code now replaces the empty value in TF state with the non-existent key, which will be sent over the wire to API.

Reply:
@chb0github Tried it but it didn't work the way we think/hope it will. So keeping the current implementation.